### PR TITLE
@W-12278342@: Loop Upgrades - Part 1: Handle for loop path with single value

### PR DESCRIPTION
**Pull request by rmohan20**
First part of a few more PRs to improve loop handling. Changes include:

1. Open up method matching for methods invoked on loop variables on class instances.
2. Minor refactor to `ApexPathExpander.java` to break down method path expansion.
3. Minor refactor to `MethodUtil.java`: extracted methods related to parameter-matching logic into a new class called `MethodTypeMatchUtil.java`.
4. Additional tests that are currently disabled.

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/graph/ops/MethodTypeMatchUtil.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static com.salesforce.graph.ops.TypeableUtil.NOT_A_MATCH;
+
+public final class MethodTypeMatchUtil {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/MethodTypeMatchUtil.java:22]: No code changes. Only extracted type-matching related methods to a new class from `MethodUtil.java`.

Comment (rmohan20) [reply sfge/src/main/java/com/salesforce/graph/ops/MethodTypeMatchUtil.java:22]: Will add javadoc.

File: sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java [hunk 2]
```diff
@@ -916,11 +692,12 @@ public static List<ApexPath> getPaths(
                                     .orElse(null);
                 }
             }
-        } else if (apexValue != null && apexValue.getDefiningType().isPresent()) {
+        } else if (apexValue != null && apexValue.getDefiningType().isPresent()) { // TODO: handle this if we have a forloopvalue of class instances
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java:695]: Todo can be removed here.

File: sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java:1]: Most of the logic remains the same.
Refactor focuses on making the method expansion more readable. I extracted multiple methods and regrouped a few things that could happen together. Will callout where relevant.

File: sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java [hunk 3]
```diff
@@ -558,199 +519,292 @@ public void visit(AbstractPathTraversalRule rule) {
...
+
+        checkMethodForRecursion(path, vertex, invocable, methodPath);
+
+
+        // Follow the path if it exists and the current path isn't the terminal method call
+
+
+        final MethodVertex method = methodPath.getMethodVertex().get();
+        final String methodDefiningType = method.getDefiningType();
+        boolean isInstanceInitializationPathAdded = false;
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java:647]: This boolean is new. Previously the same checks were made multiple times.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexForLoopValue.java [hunk 4]
```diff
@@ -153,6 +153,10 @@ public Optional<ApexValue<?>> apply(MethodCallExpressionVertex vertex, SymbolPro
 
     @Override
     public Optional<String> getDefiningType() {
+        if (!items.isEmpty()) {
+            // TODO: we should try to look at more than just the first item
+            return items.get(0).getDefiningType();
+        }
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexForLoopValue.java:159]: This change helps ApexForLoopValue have a defining type that helps this get a method match in `MethodUtil.java`.

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**